rectools/models/bert4rec.py

```diff
@@ -0,0 +1,578 @@
+            mha_output, _ = self.multi_head_attn[i](mha_input, mha_input, mha_input, need_weights=False)
+            seqs = seqs + self.dropout1[i](mha_output)
+            ff_input = self.layer_norm2[i](seqs)
+            ff_output = self.feed_forward[i](ff_input)
+            seqs = seqs + self.dropout2[i](ff_output)
+            seqs = seqs * timeline_mask
+            # seqs = self.dropout3[i](seqs)
+
+        return seqs
+
+
+# ####  --------------  Session Encoder  --------------  #### #
+
+
+class TransformerBasedSessionEncoder(torch.nn.Module):
+    """TODO"""
+
+    def __init__(
+        self,
+        n_blocks: int,
+        n_factors: int,
+        n_heads: int,
+        session_max_len: int,
+        dropout_rate: float,
+        use_pos_emb: bool = True,
+        use_causal_attn: bool = True,
+        transformer_layers_type: tp.Type[TransformerLayersBase] = BERT4RecTransformerLayers,
+        item_net_block_types: tp.Sequence[tp.Type[ItemNetBase]] = (IdEmbeddingsItemNet, CatFeaturesItemNet),
+        pos_encoding_type: tp.Type[PositionalEncodingBase] = LearnableInversePositionalEncoding,
+    ) -> None:
+        super().__init__()
+
+        self.item_model: ItemNetConstructor
+        self.pos_encoding = pos_encoding_type(use_pos_emb, session_max_len, n_factors)
+        self.emb_dropout = torch.nn.Dropout(dropout_rate)
+        self.transformer_layers = transformer_layers_type(
+            n_blocks=n_blocks,
+            n_factors=n_factors,
+            n_heads=n_heads,
+            dropout_rate=dropout_rate,
+        )
+        self.use_causal_attn = use_causal_attn
+        self.n_factors = n_factors
+        self.dropout_rate = dropout_rate
+        self.n_heads = n_heads
+
+        self.item_net_block_types = item_net_block_types
+
+    def construct_item_net(self, dataset: Dataset) -> None:
+        """TODO"""
+        self.item_model = ItemNetConstructor.from_dataset(
+            dataset, self.n_factors, self.dropout_rate, self.item_net_block_types
+        )
+
+    def encode_sessions(self, sessions: torch.Tensor, item_embs: torch.Tensor) -> torch.Tensor:
+        """
+        Pass user history through item embeddings and transformer blocks.
+
+        Returns
+        -------
+            torch.Tensor. [batch_size, session_max_len, n_factors]
+
+        """
+        session_max_len = sessions.shape[1]
+        attn_mask = None
+        if self.use_causal_attn:
+            attn_mask = ~torch.tril(
+                torch.ones((session_max_len, session_max_len), dtype=torch.bool, device=sessions.device)
+            )
+        timeline_mask = sessions != 0
+        attn_mask = ~timeline_mask.unsqueeze(1).repeat(self.n_heads, timeline_mask.squeeze(-1).shape[1], 1)
+        timeline_mask = timeline_mask.unsqueeze(-1)
+        seqs = item_embs[sessions]  # [batch_size, session_max_len, n_factors]
+        seqs = self.pos_encoding(seqs, timeline_mask)
```

Comment:
Decided to drop passing timeline_mask to pos_encoding

Reply:
fixed